stan/math/rev/mat/functor/integrate_ode_cvodes.hpp

```diff
@@ -0,0 +1,123 @@
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/rev/mat/functor/cvodes_integrator.hpp>
+#include <stan/math/rev/arr/fun/decouple_ode_states.hpp>
+#include <boost/type_traits/is_same.hpp>
+#include <ostream>
+#include <vector>
+
+namespace stan {
+  namespace math {
+    /**
+     * Return the solutions for the specified system of ordinary
+     * differential equations given the specified initial state,
+     * initial times, times of desired solution, and parameters and
+     * data, writing error and warning messages to the specified
+     * stream.
+     *
+     * This function is templated to allow the initial times to be
+     * either data or autodiff variables and the parameters to be data
+     * or autodiff variables.  The autodiff-based implementation for
+     * reverse-mode are defined in namespace <code>stan::math</code>
+     * and may be invoked via argument-dependent lookup by including
+     * their headers.
+     *
+     * The solver argument selects the solver to be used:
+     * - 0 is a non-stiff Adams-Moulton
+     * - 1 is a stiff BDF
+     * - 2 is a stiff BDF with stability limit detection
+     *
+     * @tparam F type of ODE system function.
+     * @tparam T1 type of scalars for initial values.
+     * @tparam T2 type of scalars for parameters.
+     * @param[in] f functor for the base ordinary differential equation.
+     * @param[in] y0 initial state.
+     * @param[in] t0 initial time.
+     * @param[in] ts times of the desired solutions, in strictly
+     * increasing order, all greater than the initial time.
+     * @param[in] theta parameter vector for the ODE.
+     * @param[in] x continuous data vector for the ODE.
+     * @param[in] x_int integer data vector for the ODE.
+     * @param[in] rel_tol relative tolerance of solution
+     * @param[in] abs_tol absolute tolerance of solution
+     * @param[in] max_num_steps maximal number of admissable steps
+     * between time-points
+     * @param[in] solver selects solver, admissable values 0,1,2
+     * @param[in, out] msgs the print stream for warning messages.
+     * @return a vector of states, each state being a vector of the
+     * same size as the state variable, corresponding to a time in ts.
+     */
+    template <typename F, typename T1, typename T2>
+    std::vector<std::vector<typename stan::return_type<T1, T2>::type> >
+    integrate_ode_cvodes(const F& f,
+                         const std::vector<T1>& y0,
+                         const double t0,
+                         const std::vector<double>& ts,
+                         const std::vector<T2> theta,
+                         const std::vector<double>& x,
+                         const std::vector<int>& x_int,
+                         double rel_tol = 1e-10,
+                         double abs_tol = 1e-10,
+                         long int max_num_steps = 1e8,  // NOLINT(runtime/int)
+                         const size_t solver = 1,
```

Comment:
That sounds good.

> On Apr 20, 2016, at 5:15 PM, wds15 <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/functor/integrate_ode_cvodes.hpp:
> 
> > +     * @return a vector of states, each state being a vector of the
> > +     * same size as the state variable, corresponding to a time in ts.
> > +     */
> > +    template <typename F, typename T1, typename T2>
> > +    std::vector<std::vector<typename stan::return_type<T1, T2>::type> >
> > +    integrate_ode_cvodes(const F& f,
> > +                         const std::vector<T1>& y0,
> > +                         const double t0,
> > +                         const std::vector<double>& ts,
> > +                         const std::vector<T2> theta,
> > +                         const std::vector<double>& x,
> > +                         const std::vector<int>& x_int,
> > +                         double rel_tol = 1e-10,
> > +                         double abs_tol = 1e-10,
> > +                         long int max_num_steps = 1e8,  // NOLINT(runtime/int)
> > +                         const size_t solver = 1,
> 
> So is only the solver argument to integrate_ode_cvodes a problem OR how it's done in cvodes_integrator? In any case, I would suggest for the moment to restrict things to the BDF solver only and we clean up the different solver options later.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>